src/reducers/computeRenderedColumns.js

```diff
@@ -138,6 +138,87 @@ function calculateRenderedColumnRange(state, columnAnchor) {
   };
 }
 
+/**
+ * Determine the range amd offsets of column groups to be rendered (buffer and viewport)
+ *
+ * NOTE (jordan) This alters state so it shouldn't be called
+ * without state having been cloned first.
+ *
+ * @param {!Object} state
+ * @param {{
+ *   firstIndex: number,
+ *   firstOffset: number,
+ *   lastIndex: number,
+ * }} columnAnchor
+ * @param {{
+ *   endBufferCol: number,
+ *   endViewportCol: number,
+ *   firstBufferCol: number,
+ *   firstViewportCol: number,
+ * }} columnRange
+ * @private
+ */
+function calculateRenderedColumnGroups(state, columnAnchor, columnRange) {
+  const bufferColumnCount = 0; // TODO (pradeep): calculate this similar to bufferRowCount
+  const { columnGroupBufferSet, columnGroupOffsetIntervalTree } = state;
+
+  const { availableScrollWidth, scrollableColumns, scrollableColumnGroups, columnGroupIndex } = columnWidths(state);
+  const columnCount = scrollableColumns.length;
+  const columnGroupCount = scrollableColumnGroups.length;
+
+  if (columnCount === 0 || columnGroupCount === 0) {
+    state.firstColumnGroupIndex = 0;
+    state.endColumnGroupIndex = 0;
+    state.columnGroupsToRender = [];
+    state.columnGroupOffsets = {};
+    return;
+  }
+
+  // get first and last scrollable columns in the view port
+  const startScrollableColumn = scrollableColumns[Math.max(0, columnRange.firstViewportCol)];
+  const endScrollableColumn = scrollableColumns[Math.max(0, columnRange.endViewportCol - 1)];
+
+  // now get the first and last scrollable group column's index in the view port
+  const startIdx = columnGroupIndex[startScrollableColumn.groupIdx];
+  const endIdx = columnGroupIndex[endScrollableColumn.groupIdx] + 1;
+
+  // output for this function
+  const columns = []; // state.columnsToRender
+  const columnOffsets = {}; // state.columnOffsets
+
+  // update offsets for the columns
+  let totalWidth = columnAnchor.firstOffset;
+  for (let currentIdx = startIdx; currentIdx < endIdx; currentIdx++) {
+    totalWidth += updateColumnGroupWidth(state, currentIdx);
+
+    // no need to calculate if viewport has been filled
+    if (totalWidth >= availableScrollWidth) {
+      break;
+    }
+  }
+
+  const renderedColumnsCount = endIdx - startIdx + 2 * bufferColumnCount;
+
+  // incremental way for calculating columnOffset
+  let runningOffset = columnGroupOffsetIntervalTree.sumUntil(startIdx);
+
+  // compute column index and offsets for every columns inside the viewport
+  for (let columnIdx = startIdx; columnIdx < endIdx; columnIdx++) {
```

Comment:
same logic as that of [rows/columns offsets] (https://github.com/schrodinger/fixed-data-table-2/blob/v1.0-beta/src/reducers/computeRenderedRows.js#L230)

Reply:
It seems a little weird to me that we're doing this the same way as we're doing it for rows / columns.  It seems like there's a risk of columns & column groups becoming inconsistent / out of sync.

Would it be better to use our knowledge of column widths and the final scroll X and then have a selector compute the column groups on screen from that information?

Reply:
> It seems a little weird to me that we're doing this the same way as we're doing it for rows / columns. It seems like there's a risk of columns & column groups becoming inconsistent / out of sync.
> 
> Would it be better to use our knowledge of column widths and the final scroll X and then have a selector compute the column groups on screen from that information?

Not sure if I got you here... We could just find the range of column groups to be displayed through the selector and then just use it at our views, but the aim of this loop/reducer is that
* offsets can be modified efficiently
* integer buffer set minimizes cell unmounts for grouped header

point 1 doesn't really work since we prefetch all columns unlike rows :| (related to #464))
but point 2 definitely stands especially when cells are heavy

Reply:
Ah okay, so the big gain here is to avoid rearranging column groups as much as possible.  Would it be possible to keep that aspect, but remove the "column group anchor" parts, and just get that from the scrollX / column anchor parts?  I don't see a reason why we need

Reply:
Ah okay, so the big gain here is to avoid rearranging column groups as much as possible. 

I read this as using "column group anchors" to determine what should be rendered and how it should be offset.  Is that part of this logic too, or was I just misunderstanding the use of things like firstColumnGroupIndex, endColumnGroupIndex, etc

Reply:
Ah no, I should have explained better: there's no column group anchor.

We only keep 1 source for scrolling, which is the column anchor for our regular columns. 
We then use the column anchor to find `firstColumnIndex` and `endColumnIndex`, which will further be used to find `firstColumnGroupIndex` and `endColumnGroupIndex`.

Reply:
ah okay.  I misunderstood this.  Thanks for the clarification.